eclass/tests/zig-utils.sh

```diff
@@ -0,0 +1,311 @@
+	_EOF_
+
+	zig_test_compile_and_run() {
+		eindent
+		einfo "${1}: ${2}"
+		eoutdent
+
+		if [[ "${2}" == native* ]]; then
+			return 0
+		fi
+
+		if [[ "${1}" == arm* ]]; then
+			# Bunch of unimplemented Zig
+			# routines for more "bare-bone" arm.
+			case "${2}" in
+				# Some errors in inline assembly, likely too low baseline
+				generic-soft_float | generic+soft_float) return 0;;
+				# undefined symbol: __sync_fetch_and_add_1
+				# compiler-rt not implemented in upstream for it yet:
+				# https://github.com/ziglang/zig/issues/4959
+				generic+v5te-soft_float) return 0;;
+				*) ;;
+			esac
+		fi
+
+		local ret=0
+
+		local args=(
+			-target "${1}"
+			-mcpu "${2//:no-run/}"
+			main.zig
+		)
+		if ! zig build-exe "${args[@]}" > /dev/null; then
+			eerror "Failed to compile for: ${1}, ${2}"
+			((++ret))
+			return ${ret}
+		fi
+
+		# Can't run macOS binaries in Linux user-mode QEMU emulators
+		if [[ "${1}" == *macos* || "${2}" == *":no-run" ]]; then
+			return ${ret}
+		fi
+
+		if ! ./main > /dev/null; then
+			eerror "Failed to run for: ${1}, ${2}"
+			((++ret))
+		fi
+
+		return ${ret}
+	}
+else
+	zig_test_compile_and_run() { :; }
+fi
+
+test-convert_c_env_to_zig_tuple() {
+	local ret=0
+
+	local -n map=${1}
+
+	local c_tuple
+	for key in "${!map[@]}"; do
+		local expected="${map["${key}"]}"
+
+		local c_tuple
+		local c_flags
+		if [[ "${key}" == *:* ]]; then
+			c_tuple="${key%%:*}"
+			c_flags="${key##*:}"
+		else
+			c_tuple="${key}"
+			c_flags=""
+		fi
+
+		local actual="$(zig-utils_c_env_to_zig_target "${c_tuple}" "${c_flags}")"
+		if [[ "${expected}" != "${actual}" ]]; then
+			eerror "Translating ${c_tuple}: expected ${expected}, found ${actual}"
+			((++ret))
+			continue
+		fi
+
+		local zig_cpu="$(zig-utils_c_env_to_zig_cpu "${c_tuple}" "${c_flags}")"
+		if ! zig_test_compile_and_run "${expected}" "${zig_cpu}"; then
+			((++ret))
+			continue
+		fi
+	done
+
+	return ${ret}
+}
+
+test-convert_c_env_to_zig_cpu() {
+	local ret=0
+
+	local -n map=${1}
+	local chost=${2}
+
+	local c_flags
+	for c_flags in "${!map[@]}"; do
+		local expected="${map["${c_flags}"]}"
+		local actual="$(zig-utils_c_env_to_zig_cpu "${chost}" "${c_flags}")"
+		if [[ "${expected//:no-run/}" != "${actual}" ]]; then
+			eerror "Translating ${c_flags}: expected ${expected//:no-run/}, found ${actual}"
+			((++ret))
+			continue
+		fi
+
+		local zig_target="$(zig-utils_c_env_to_zig_target "${chost}" "${c_flags}")"
+		if ! zig_test_compile_and_run "${zig_target}" "${expected}"; then
+			((++ret))
+			continue
+		fi
+	done
+
+	return ${ret}
+}
+
+tbegin '"C tuple to Zig tuple"'
+declare -A c_to_zig_map=(
+	# Just remove "vendor" field
+	[aarch64-unknown-linux-gnu]=aarch64-linux-gnu
+	[arm-unknown-linux-musleabi]=arm-linux-musleabi
+	[x86_64-pc-linux-gnu]=x86_64-linux-gnu
+	[loongarch64-unknown-linux-gnu]=loongarch64-linux-gnu
+	[powerpc64le-unknown-linux-gnu]=powerpc64le-linux-gnu
+
+	# ARM big-endian
+	[armbe-unknown-linux-gnueabi]=armeb-linux-gnueabi
```

Comment:
Do we actually have "armbe" tuples? I've only ever seen "arm*eb".

Reply:
Yeah, it's invalid:

```
$ clang --target=armbe-pc-linux-gnu -x c /dev/null
clang: error: unknown target triple 'armbe-pc-linux-gnu'
```

Reply:
Changed to real-life example from https://bugs.gentoo.org/924920

2d7c12de22bb